src/vision/status_quo/grace_wants_to_integrate_c_api.md

```diff
@@ -0,0 +1,176 @@
+is tempted to write a callback-based architecture by relying on the blocking interface that waits; however, early the
+next morning the customer comes back and informs her that they are scaling up the system, and that there will now be 5
+cameras instead of 1.
+
+She knows from experience that she cannot rely on having 5 threads blocking just for getting camera frames, because the
+embedded system she is deploying to only has 2 cores total! Her team would be introducing a lot of overhead into the
+system with the continuous context switching of every thread.  Fortunately, Grace notices the similarities between the
+polling interface in the underlying C library and the `Poll` type returned by Rust's `Future` trait. "Surely," she
+thinks, "I can asynchronously interleave polls to each camera over a single thread, and process frames as they become
+available!" Such a thing would be quite difficult in C while guaranteeing memory safety was maintained. However, Grace's
+team has already dodged that bullet thanks to writing a thin wrapper in Rust that manages these tricky lifetimes!
+
+### The first problem: polls and wake-ups
+
+Grace sets out to start writing the pipeline to get frames from the cameras. She realizes that while the polling call
+that the manufacturer provided in their library is similar in nature to a future, it doesn't quite encompass everything.
+In C, one might have to set some kind of heartbeat timer for polling. Grace explains to her team that this heartbeat is
+similar to how the `Waker` object works in a `Future`'s `Context` type, in that it is how often the execution
+environment should re-try the future if the call to `poll` returns `Poll::Pending`.
+
+Something Grace learns to love immediately about Rust is that writing the `Future` here does not require her team to
+write their own execution environment. In fact, the future can be entirely written independent of the execution
+environment. She quickly writes an async method to represent the polling process:
+
+```rust
+/// Gets the next frame from the camera, waiting `retry_after` time until polling again if it fails.
+///
+/// Returns Some(frame) if a frame is found, or None if the camera is disconnected or goes down before a frame is
+/// available.
+async fn next_frame(camera: &Camera, retry_after: Duration) -> Option<Frame> {
+    while camera.is_available() {
+        if let Some(frame) = camera.poll() {
+            return Some(frame);
+        } else {
+            task::sleep_for(retry_after).await;
+        }
+    }
+
+    None
+}
+```
+
+The underlying C API doesn't provide any hooks that can be used to wake the `Waker` object on this future up, so Grace
+and her team decide that it is probably best if they just choose a sufficiently balanced `retry_after` period in which
+to try again. It does feel somewhat unsatisfying, as calling `sleep_for` feels about as hacky as calling
```

Comment:
I sympathize with Grace :)

Just a small personal anecdote:  Tokio as a [tutorial](https://tokio.rs/tokio/tutorial/async) that covers how to write your own custom `Future` that will wait for some interval to elapse.  The future that tokio has you write simply spawns a blocking `sleep` in a new thread.  When I first saw this I felt the same type of "unsatisfaction" that Grace probably felt too.

Reply:
Thank you for the feedback! I did not know that tutorial existed, and actually I've mostly stuck to async-std over tokio, so I must have missed a lot of valuable tutorials!

I admit, this is a really weird part of async that feels unsatisfying, but it's actually not that bad with regards to the runtime. At least, it isn't as bad as doing the equivalent with `std::thread::sleep` on a thread (which is usually a sign you want a condition variable!).

Reply:
Do you think Grace went looking for a tutorial on how to implement a custom Future, and was disappointed when she didn't find one?   Or do you think Grace read the API docs for Future and they made enough sense to her that she felt comfortable jumping in and trying a custom implementation?

Reply:
Certainly the latter in this case. Grace definitely read the docs for the `Future` trait, and then intuitively understood how async-await de-sugars to a custom future. 

Grace may have read [this article on why async trait fns are hard](https://smallcultfollowing.com/babysteps/blog/2019/10/26/async-fn-in-traits-are-hard/) which provides a ton of background for intuiting the async-await and how the compiler structures / de-sugars that syntax, but otherwise didn't feel any additional need for tutorials given that plus past experience + the rustdocs. The loop -> sleep cycle seemed natural, albeit did feel unidiomatic or unsatisfying coming from C or C++, where that would have probably been a code-smell.

Reply:
On that note, is this worth adding to the story? The blog post is fantastic, and a great resource for people jumping into async from C/C++, but I don't know if it is directly applicable without feeling like I forced it in somehow...

Reply:
I don't want to unduly color Grace's experiences here, but maybe adding a small note about this into the story would be good.  We have a lot of stories about users who get stuck in various ways on simple things, so it might be nice to mention that some characters (like Grace here) already have the right experience and intuition to jump directly into things like writing custom futures (even when being relatively new to async rust)

Reply:
I definitely think it'd be good to elaborate on this part.